### PR TITLE
Refactoring

### PR DESCRIPTION
**Pull request by jerpint**

### CODE REVIEW

File: buster/chatbot.py [hunk 1]
```diff
@@ -32,7 +32,7 @@ class ChatbotConfig:
     embedding_model: OpenAI model to use to get embeddings.
     top_k: Max number of documents to retrieve, ordered by cosine similarity
     thresh: threshold for cosine similarity to be considered
-    max_chars: maximum number of characters the retrieved documents can be. Will truncate otherwise.
+    max_words: maximum number of words the retrieved documents can be. Will truncate otherwise.
```
Comment (marondeau-mila) [inline buster/chatbot.py:35]: Is it words or token ?

File: buster/chatbot.py [hunk 2]
```diff
@@ -78,25 +80,24 @@ def _init_documents(self):
         logger.info(f"embeddings loaded.")
 
     def _init_unk_embedding(self):
-        logger.info("Generating UNK token...")
```
Comment (marondeau-mila) [inline buster/chatbot.py:81]: I don't like magic function in a constructor. I would suggest using a dataclass, with a factory method in the config where this function is called.

File: buster/chatbot.py [hunk 3]
```diff
@@ -121,59 +122,70 @@ def rank_documents(
...
-            logger.info(f"GPT Response:\n{response_text}")
-            return response_text
+            return openai.Completion.create(**completion_kwargs)
 
         except Exception as e:
             # log the error and return a generic response instead.
             import traceback
 
             logger.error("Error connecting to OpenAI API")
             logging.error(traceback.format_exc())
```
Comment (marondeau-mila) [inline buster/chatbot.py:166]: 1. Do not do an `except Exception`. Catch the ones that are truly temporary failures.
2. `logger.exception` should be better.

**Review by marondeau-mila — COMMENTED** [review]
I'm not a big fan of creating class methods to deal with test. Actually solving the problem (mocking the class in the test) would be the solution.

File: buster/chatbot.py (hunk 1 above)
Comment (jerpint) [reply buster/chatbot.py:35]: right now it's implemented as words, with the rough guesstimate that 4000 tokens limit ~= 3000 words. We have an open issue to start converting everything to proper token counts, shouldn't be too hard just not in the scope of this PR

File: buster/chatbot.py (hunk 3 above)
Comment (jerpint) [reply buster/chatbot.py:166]: that's the thing, I'm not sure which exceptions we can catch, openAI's API is a bit sparse and I've been catching different ones as i've been debugging

good call on the `logger.exception`, fixing now

File: buster/chatbot.py (hunk 2 above)
Comment (jerpint) [reply buster/chatbot.py:81]: I agree it's a little clunky, but this method will probably evolve over time, i'd rather not spend too much time factoring it right now as it'll likely become more important in the future

**Comment by jerpint** [comment]
@marondeau-mila are you suggesting simply removing all `classmethod` decorators and having the functions be native instead?

**Comment by marondeau-mila** [comment]
The reason you added the classmethods was for testing. I'm saying that's a hack. The solution is to use mocking to test the methods directly. That is, modify the test to fit the code, not the code to fit the test.

File: buster/chatbot.py [hunk 4]
```diff
@@ -162,8 +162,8 @@ def get_gpt_response(cls, **completion_kwargs):
             # log the error and return a generic response instead.
             import traceback
 
-            logger.error("Error connecting to OpenAI API")
-            logging.error(traceback.format_exc())
+            logger.exception("Error connecting to OpenAI API")
```
[the remark this answers is not in the document]
Comment (jerpint) [reply buster/chatbot.py]: you are correct, that was unnecessary, the error logging is now handled directly with the single logger

File: buster/chatbot.py [hunk 5]
```diff
@@ -160,10 +160,7 @@ def get_gpt_response(cls, **completion_kwargs):
 
         except Exception as e:
             # log the error and return a generic response instead.
-            import traceback
-
-            logger.exception("Error connecting to OpenAI API")
-            logging.exception(traceback.format_exc())
+            logger.exception("Error connecting to OpenAI API. See traceback: %s", e)
```
Comment (marondeau-mila) [inline buster/chatbot.py:163]: The traceback is automatically added to the message. You don't need to append it.

Comment (jerpint) [reply buster/chatbot.py]: nice, you are right, i didn't know that! updated

File: buster/chatbot.py [hunk 6]
```diff
@@ -192,25 +197,53 @@ def add_sources(self, response: str, matched_documents: pd.DataFrame):
...
+        """
+        response_embedding = get_embedding(
+            response,
+            engine=engine,
+        )
+        score = cosine_similarity(response_embedding, unk_embedding)
+        logger.info(f"UNK score: {score}")
+
+        # Liekly that the answer is meaningful, add the top sources
+        return True if score < unk_threshold else False
```
Comment (hbertrand) [inline buster/chatbot.py:218]: `return score < unk_threshold`

File: buster/chatbot.py [hunk 7]
```diff
@@ -121,59 +121,64 @@ def rank_documents(
...
 
     def generate_response(self, prompt: str, matched_documents: pd.DataFrame) -> str:
         """
         Generate a response based on the retrieved documents.
         """
         if len(matched_documents) == 0:
             # No matching documents were retrieved, return
-            response_text = "I did not find any relevant documentation related to your question."
-            return response_text
+            return "I did not find any relevant documentation related to your question."
```
Comment (hbertrand) [inline buster/chatbot.py:170]: Seems redundant since it's also in format_response

**Review by hbertrand — CHANGES_REQUESTED** [review]

File: buster/chatbot.py (hunk 7 above)
Comment (jerpint) [reply buster/chatbot.py:170]: hmm youre right, it shouldnt be in both, let me fix that

Comment (jerpint) [reply buster/chatbot.py]: i just changed it to something more "clever". If the length of documents is 0, we will instead return the `unknown_prompt` text, which will then always trigger the UNK token in the `is_relevant` function, bypassing the need to check for the length.

File: buster/chatbot.py (hunk 6 above)
Comment (jerpint) [reply buster/chatbot.py]: fixed, good catch

**Review by hbertrand — APPROVED** [review]

**Pull request merged**